### PR TITLE
Add qlf_k6n10 plugin

### PR DESCRIPTION
**Pull request by samycharas**
Following the recent addition of qlf_k4n8 plugin, this PR adds qlf_k6n10 synthesis flow.
It is required for the upcoming pull request for qlf_k6n10 device on [symbiflow-arch-defs](https://github.com/SymbiFlow/symbiflow-arch-defs).
Instead of creating 2 directories for each qlf plugin, I've merged them into ql-qlf-plugin. 
This required to use another Makefile_test.common ( under ql-qlf-plugin/common/utils folder ) file to match path difference created by an additional parent folder
I also replaced the verilog source file location to a common directory under ql-qlf-plugin folder to avoid file duplication.

Side note : I'm looking to enable the BRAM test in an upcoming PR.

Signed-off-by: Samy Charas samy.charas@gmail.com

### CODE REVIEW

**Comment by samycharas** [comment]
@lnsharma FYI

**Review by lnsharma — COMMENTED** [review]
synth_quicklogic.cc should be common to all QuickLogic devices so it should be placed under a device folder. Currently I see it in ql-qlf-k4n8-plugin/synth_quicklogic.cc
This file should be common and its command line option '-family' can accept the supported devices.

File: ql-qlf-plugin/common/utils/Makefile_test.common
Comment (lnsharma) [inline ql-qlf-plugin/common/utils/Makefile_test.common:17]: Instead use conditional assignment for these variables.

File: ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc
Comment (lnsharma) [inline ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc:1]: Do not add synth_quicklogic.cc separately for different devices. Instead add only synth_quicklogic.cc which support both the devices.

**Comment by tpagarani** [comment]
@kgugala, would you be able to review this PR?

File: ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc
Comment (tpagarani) [inline ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc:221]: Does the K6N10 architecture support DLATCH? There is no techmap specified for DLATCH in the flow.

File: ql-qlf-plugin/ql-qlf-k6n10-plugin/cells_sim.v
Comment (tpagarani) [inline ql-qlf-plugin/ql-qlf-k6n10-plugin/cells_sim.v:24]: There is gclkbuff as part of maps. Is this redundant?

**Review by kgugala — CHANGES_REQUESTED** [review]
The biggest problem with this code is that it does not follow the way the rest of the plugins are implemented.

You should not copy global Makefiles. Global Makefiles will automatically pick plugins and their tests. Plugins should simply include common Makefiles.

See comments in https://github.com/SymbiFlow/yosys-symbiflow-plugins/blob/master/Makefile_plugin.common#L2 and https://github.com/SymbiFlow/yosys-symbiflow-plugins/blob/master/Makefile_test.common#L1

Also, you can take a look at https://github.com/SymbiFlow/yosys-symbiflow-plugins/pull/85 and see how the qlf plugin was originally added

**Comment by tpagarani** [comment]
@lnsharma @kgugala,  i see some confusion around proposal for plugin. I think lalit is suggesting to create 2 different plugin (one .so per family), however he is proposing to keep synth_quicklogic.cc common. Whereas what Karol is proposing that there should be single plugin (a single .so) handling both families.  Is that correct understanding I have?

File: ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc
Comment (lnsharma) [reply ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc:1]: > @lnsharma @kgugala, i see some confusion around proposal for plugin. I think lalit is suggesting to create 2 different plugin (one .so per family), however he is proposing to keep synth_quicklogic.cc common. Whereas what Karol is proposing that there should be single plugin (a single .so) handling both families. Is that correct understanding I have?

If we have one synth_quicklogic.cc, then it is obvious that there is going to be one .so. Correct me if I am wrong. That is what I have been saying to have one synth_quicklogic.cc.

Comment (tpagarani) [reply ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc:1]: > > @lnsharma @kgugala, i see some confusion around proposal for plugin. I think lalit is suggesting to create 2 different plugin (one .so per family), however he is proposing to keep synth_quicklogic.cc common. Whereas what Karol is proposing that there should be single plugin (a single .so) handling both families. Is that correct understanding I have?
> 
> If we have one synth_quicklogic.cc, then it is obvious that there is going to be one .so. Correct me if I am wrong. That is what I have been saying to have one synth_quicklogic.cc.

Ok. Thanks for the clarifying this.

File: ql-qlf-plugin/common/utils/Makefile_test.common
Comment (samycharas) [reply ql-qlf-plugin/common/utils/Makefile_test.common]: I modified the plugin folder structure to align on other plugins this time, we do not need modification to this anymore.

File: ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc
Comment (samycharas) [reply ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc]: I've merged the 2 synth_quicklogic.cc, we now have one plugin for both family.

File: ql-qlf-plugin/ql-qlf-k6n10-plugin/cells_sim.v
Comment (samycharas) [reply ql-qlf-plugin/ql-qlf-k6n10-plugin/cells_sim.v]: Agreed, I moved this file to a common folder.

File: ql-qlf-plugin/ql-qlf-k6n10-plugin/tests/ql_qlf_k6n10/run-test.sh
Comment (samycharas) [reply ql-qlf-plugin/ql-qlf-k6n10-plugin/tests/ql_qlf_k6n10/run-test.sh]: I removed the script for now, @lnsharma do you need this script?

File: ql-qlf-plugin/common/verilog/latch.v
Comment (samycharas) [reply ql-qlf-plugin/common/verilog/latch.v]: Corrected, it's now added to latches.v file

File: ql-qlf-plugin/common/verilog/dff.v
Comment (samycharas) [reply ql-qlf-plugin/common/verilog/dff.v]: Corrected, it's now added dffs.v file

File: ql-qlf-plugin/Makefile
Comment (samycharas) [reply ql-qlf-plugin/Makefile]: I've aligned on other plugin folder structure.

File: ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc
Comment (samycharas) [reply ql-qlf-plugin/ql-qlf-k6n10-plugin/synth_quicklogic.cc]: Architecture doesn't support DLATCH, I'll be removing this.

**Comment by samycharas** [comment]
Hi @kgugala, I've modified the pull request to align this plugin on plugins already present in the directory. Could you please review it again?

**Comment by kgugala** [comment]
> Hi @kgugala, I've modified the pull request to align this plugin on plugins already present in the directory. Could you please review it again?

thanks @samycharas, the PR looks better now. There are still conflicts. Can you rebase the PR on top of master?

**Comment by samycharas** [comment]
I've rebase on top of the master, CI is now running, so I assume there should not be conflict anymore.

**Comment by kgugala** [comment]
can you squash commits in a way we don't have git conflict artifacts in any commit? Leaving them will make any future bisect really hard.

**Comment by samycharas** [comment]
I've rebased and left commit for :
- qlf_k6n10 plugin.
- Merge of qlf_k4n8 & qlf_k6n10 plugin.
Does this looks better to you now?

**Comment by samycharas** [comment]
@kgugala It does look like CI is failing due to a package server being unstable, should I push a dummy commit to restart CI, or could you please rerun CI if you have access to that ? It should pass without the package server timeout.

**Comment by kgugala** [comment]
> @kgugala It does look like CI is failing due to a package server being unstable, should I push a dummy commit to restart CI, or could you please rerun CI if you have access to that ? It should pass without the package server timeout.

I have restarted the build

File: ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v
Comment (kgugala) [inline ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v:96]: formatting is a bit off here

File: ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v
Comment (kgugala) [inline ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v:102]: formatting

File: ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v
Comment (kgugala) [inline ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v:113]: format

File: ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v
Comment (kgugala) [inline ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v:126]: format

File: ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v
Comment (kgugala) [inline ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v:150]: format

File: ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v
Comment (kgugala) [inline ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v:194]: format

**Review by kgugala — CHANGES_REQUESTED** [review]
LGTM, some minor formatting comments - once fixed we can land this PR

File: ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v
Comment (samycharas) [reply ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v]: Fixed the formatting

File: ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v
Comment (samycharas) [reply ql-qlf-plugin/ql-qlf-k6n10/qlf_k6n10_cells_sim.v]: Fixed the formatting

**Review by kgugala — APPROVED** [review]

**Comment by kgugala** [comment]
Looks good, let's wait for CI

**Pull request merged**
